mitodl/mit-open
# create api routes for user subscribe/unsubscribe to search

### What are the relevant tickets?

Closes #755 


### Description (What does it do?)
This PR introduces a "user subscription" (/api/v1/learning_resources_search_subscription) endpoint that allows for the following:
1. Subscribing a user to a search query via POST
2. Unsubscribing a user from a query via DELETE
3. Checking the subscription status for a user via GET

### How can this be tested?
How I have been testing this on the frontend is by going to the search page http://od.odl.local:8063/search/?q=testing and then opening up dev-tools and copying the json request "as fetch" and then modifying it for the learning_resources_search_subscription route. 

My fetch requests look like so. you should be able to re-use by adding a valid csrf token (run these in developer tools console):

**Subscribe user**
<pre><code>
fetch("http://od.odl.local:8063/api/v1/learning_resources_search_subscribe/", {
  "headers": {
    "accept": "application/json, text/plain, */*",
    "accept-language": "en-US,en;q=0.9",
    "Content-Type": "application/json",
    "x-csrftoken":  <b>"csrf-token"</b>
  },
  "body": JSON.stringify({
  "aggregations":[
    "resource_type",
    "topic",
    "offered_by"
  ],
  "limit":"10",
  "offset":"0",
  "q":"testing",
  "resource_type":[
    "course",
    "program",
    "podcast"
  ]
}),
  "method": "POST",
  "mode": "cors",
  "credentials": "include"
});
</code></pre>

**Unsubscribe user**
<pre>
<code>

fetch("http://od.odl.local:8063/api/v1/learning_resources_search_subscribe/", {
  "headers": {
    "accept": "application/json, text/plain, */*",
    "accept-language": "en-US,en;q=0.9",
    "Content-Type": "application/json",
    "cache-control": "no-cache",
    "pragma": "no-cache",
    "x-csrftoken": <b>"csrf-token"</b>
  },
  "referrer": "http://od.odl.local:8063/search/?q=testing",
  "referrerPolicy": "same-origin",
  "body": JSON.stringify({
  "aggregations":[
    "resource_type",
    "topic",
    "offered_by"
  ],
  "limit":"10",
  "offset":"0",
  "q":"testing",
  "resource_type":[
    "course",
    "program",
    "podcast"
  ]
}),
  "method": "DELETE",
  "mode": "cors",
  "credentials": "include"
});


</code>
</pre>

You can then check the subscribed/unsubscribed status by visiting the get request for the endpoint (http://od.odl.local:8063/api/v1/learning_resources_search_subscribe/?aggregations=resource_type&aggregations=topic&aggregations=offered_by&limit=10&offset=0&q=testing&resource_type=course&resource_type=program&resource_type=podcast)

### Additional Notes
When storing the original query we strip 'limit', 'sortby' and 'offset' which are not relevant to the query - for example 
```
{
'query': 'test',
'limit': 20,
'offset': 5
}
```
 and 
```{
'query': 'test',
'limit': 100,
'sortby': 'title',
}
```
should resolve to the same query ```{'query': 'test'}```
